apps/sms/views/shared/js/navigation.js

```diff
@@ -1,361 +1,726 @@
-/* global Promise,
-      EventDispatcher
-*/
-
-/* exported Navigation */
+/*global
+  Utils
+ */
+(function(exports) {
 
 'use strict';
 
-(function(window) {
+/**
+ * The Defer type is useful when creating promises.
+ * @typedef {Object} Defer
+ * @property {function(*)} resolve The Promise's resolve function.
+ * @property {function(*)} reject The Promise's reject function.
+ * @property {Promise} promise The actual promise.
+ */
+
+/**
+ * @function
+ * The debug function translates to console.log in debug mode.
+ *
+ * @param {...*} args The values to be displayed to the console.
+ */
+const debug = 1 ? (...args) => console.log(...args) : () => {};
+
+/**
+  * This contains an array of view characteristics, necessary to handle
+  * navigation requests.
+  * Each item has the following properties:
+  * - name: this is the name of the view. This will be used also in the URL
+  *   hash if the same HTML file holds several views.
+  * - url: this is the URL used to match the current location and find the
+  *   current view.
+  * - view: this is the actual JavaScript object that manages this view.
+  * - previous (optional): this is the name of the previous view; this is
```

Comment:
>> Maybe we could record the panel changes and use that instead, and use "previous" only when we don't know the previous panel (for example because the user accessed a view directly and presses the "back" button that uses toPanel ?).

> Mmm, maybe, kind of default back route would work :)

Yeah I agree default back behavior should be able to handle this. But just wondering do we really need to migrate this into message app. For example, if user send the new message and entered conversation view, would he/she expect to see the new message view again when back button pressed? We might need to have more discuss with UX about the navigation behavior in the future.